lib/srv/alpnproxy/proxy.go

```diff
@@ -0,0 +1,343 @@
+		return nil, trace.Wrap(err)
+	}
+	var supportedProtocols []string
+	for k := range cfg.Router.alpnHandlers {
+		supportedProtocols = append(supportedProtocols, k)
+	}
+	return &Proxy{
+		cfg:                cfg,
+		log:                cfg.Log,
+		supportedProtocols: supportedProtocols,
+	}, nil
+}
+
+// Serve starts accepting connections.
+func (p *Proxy) Serve(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	p.cancel = cancel
+	p.cfg.TLSConfig.NextProtos = p.supportedProtocols
+	for {
+		clientConn, err := p.cfg.Listener.Accept()
+		if err != nil {
+			if utils.IsOKNetworkError(err) || trace.IsConnectionProblem(err) {
+				return nil
+			}
+			return trace.Wrap(err)
+		}
+		go func() {
+			if err := p.handleConn(ctx, clientConn); err != nil {
+				if err := clientConn.Close(); err != nil {
+					p.log.WithError(err).Warnf("failed to close client connection")
+				}
+				p.log.WithError(err).Warnf("failed to handle client connection")
+			}
+		}()
+	}
+}
+
+func (p *Proxy) handleConn(ctx context.Context, clientConn net.Conn) error {
+	hello, conn, err := p.readHelloMessageWithoutTLSTermination(clientConn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	handlerDesc, err := p.getHandlerDescBaseOnClientHelloMsg(hello)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if handlerDesc.ForwardTLS {
+		return trace.Wrap(handlerDesc.Handler(ctx, conn))
+	}
+
+	tlsConn := tls.Server(conn, p.cfg.TLSConfig)
+	if err := tlsConn.SetReadDeadline(p.cfg.Clock.Now().Add(p.cfg.ReadDeadline)); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := tlsConn.Handshake(); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := tlsConn.SetReadDeadline(time.Time{}); err != nil {
+		return trace.Wrap(err)
+	}
+
+	isDatabaseConnection, err := utils.IsDatabaseConnection(tlsConn.ConnectionState())
+	if err != nil {
+		p.log.WithError(err).Debug("Failed to check if connection is database connection.")
+	}
+	if isDatabaseConnection {
+		return trace.Wrap(p.handleDatabaseConnection(ctx, tlsConn))
+	}
+	return trace.Wrap(handlerDesc.Handler(ctx, tlsConn))
+}
+
+// readHelloMessageWithoutTLSTermination allows reading a ClientHelloInfo message without termination of
+// incoming TLS connection. After calling readHelloMessageWithoutTLSTermination function a returned
+// net.Conn should be used for further operation.
+func (p *Proxy) readHelloMessageWithoutTLSTermination(conn net.Conn) (*tls.ClientHelloInfo, net.Conn, error) {
+	buff := new(bytes.Buffer)
+	var hello *tls.ClientHelloInfo
+	tlsConn := tls.Server(readOnlyConn{reader: io.TeeReader(conn, buff)}, &tls.Config{
+		GetConfigForClient: func(info *tls.ClientHelloInfo) (*tls.Config, error) {
+			hello = info
+			return nil, nil
+		},
+	})
+	if err := conn.SetReadDeadline(p.cfg.Clock.Now().Add(p.cfg.ReadDeadline)); err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	// Following TLS handshake fails on the server side with error: "no certificates configured" after server
```

Comment:
This functionality is invoked only to read the Hello message  without terminating TLS connection so the flow looks like:
1) Client sends hello message.
2) "Fake" TLS handshake handler reads hello message using readOnlyConn without TLS termination.
3) Fake TLS handler creates buffered connection, bytes read during fake TLS Handshake are prepended to connection. 
4) Fake TLS handler returns a bufferedConn that is used for further processing.
5) Routing connection based on hello message and "real" TLS termination.  

The server TLS config has only configured `GetConfigForClient` func without providing any Certificates. 
https://github.com/gravitational/teleport/pull/7524/files/f67390f90e1c3e32e69c9e78a2854b9e2277a613#diff-449a76d8d1fc49a989f45a0978b522141ac0da501a115d403dc1f0c2c22e05f1R232

The code doesn't relays on any particular TLS error but if server cert are missing the handshake must fails.

Reply:
> Can this failed fake handshake somehow affect the client and/or the future "real" handshake?

Thus for fake TLS handshake we are using readonly connection this "fake" handshake should not affect any further logic. 

> Can your fake tlsConn return a proper tls.Config so this handshake succeeds also? We have p.cfg.TLSConfig here?

I don't want to terminate the TLS traffic here. If `handlerDesc.ForwardTLS` is set to true (for example case of k8 handler) we need to forward the connection to the handler without TLS termination because k8 handler uses different certs.